### ISSUE TITLE
Dropping types while joining Unions with Any

### ISSUE THREAD

**Issue by hauntsaninja** [python/mypy#9425]
The revealed type in `maybenotokay` is probably a bug, especially given the revealed type in `alsookay`.
Dropping Any in `defnotokay` seems definitely not okay.

```
from typing import Any, Union

class A: ...
class B: ...
class C: ...
class D: ...

def okay(x: Union[A, B], y: Union[C, D], b: bool):
    reveal_type(x if b else y)  # Union[main.A, main.B, main.C, main.D]

def maybenotokay(x: Union[A, B], y: Union[Any, D], b: bool):
    reveal_type(x if b else y)  # Union[Any, main.D]
    
def alsookay(x: Union[A, Any], y: Union[C, D], b: bool):
    reveal_type(x if b else y)  # Union[main.A, Any, main.C, main.D]

def defnotokay(x: Union[A, Any], y: Union[A, D], b: bool):
    reveal_type(x if b else y)  # Union[main.A, main.D]
```

Thanks to an example in https://dimvar.blogspot.com/2020/09/trying-out-mypy.html

**Comment by changpotato** [comment]
Hi, can I contribute to this? I've never contributed to mypy before, but this issue seems interesting. If I can, please let me know roughly which part I should start digging into!

**Comment by hauntsaninja** [comment]
Please do! I haven't looked into it (and don't have time to provide good guidance), but naively, this is an issue with joins, so looking at join.py seems like a good idea (the mypy wiki has a description of what a join is). Typically the best thing to do with mypy is add breakpoints in interesting places (such as wherever we reveal_type) and work backwards to trace code flow.

**Comment by momohatt** [comment]
Isn't this already fixed in #9301 though?
If you test the above program on current master you will get `Union[main.A, main.B, Any, main.D]` for `maybenotokay` and `Union[main.A, Any, main.D]` for `defnotokay`, both of which seem fine.

**Comment by hauntsaninja** [comment]
So it is! Thanks :-)
